src/hevm/src/EVM/UnitTest.hs

```diff
@@ -422,15 +444,122 @@ runTest opts@UnitTestOptions{..} vm (ConcreteTest testName, types) = liftIO $ ca
+       let cs = view (env . contracts) vm
+           noCode c = case view contractcode c of
+             RuntimeCode c' -> len c' == 0
+             _ -> False
+           mutable m = view methodMutability m `elem` [NonPayable, Payable]
+           knownAbis :: Map Addr SolcContract
+           knownAbis =
+             -- exclude contracts without code
+             Map.filter (not . BS.null . view runtimeCode) $ 
+             -- exclude contracts without state changing functions
+             Map.filter (not . null . Map.filter mutable . view abiMap) $ 
+             -- exclude testing abis
+             Map.filter (isNothing . preview (abiMap . ix unitTestMarkerAbi)) $
+             -- pick all contracts with known compiler artifacts
+             fmap fromJust (Map.filter isJust $ Map.fromList [(addr, lookupCode (view contractcode c) dapp) | (addr, c)  <- Map.toList $ cs])
+       -- go to IO and generate a random valid call to any known contract
+       liftIO $ do
+         -- select random contract
+         (target, solcInfo) <- generate $ elements (Map.toList knownAbis)
+         -- choose a random mutable method
+         (_, (Method _ inputs sig _ _)) <- generate (elements $ Map.toList $ Map.filter mutable $ view abiMap solcInfo)
+         let types = snd <$> inputs
+         -- set the caller to a random address with 90% probability, 10% known EOA address
+         let knownEOAs = Map.keys $ Map.filter noCode cs
+         AbiAddress caller' <-
+           if null knownEOAs
+           then generate $ genAbiValue AbiAddressType
+           else generate $ frequency
+             [ (90, genAbiValue AbiAddressType)
+             , (10, AbiAddress <$> elements knownEOAs)
+             ]
+         -- make a call with random valid data to the function
+         args <- generate $ genAbiValue (AbiTupleType $ Vector.fromList types)
+         let cd = abiMethod (sig <> "(" <> intercalate "," ((pack . show) <$> types) <> ")") args
+         -- increment timestamp with random amount
+         timepassed <- num <$> generate (arbitrarySizedNatural :: Gen Word32)
+         let ts = fromMaybe (error "symbolic timestamp not supported here") $ maybeLitWord $ view (block . timestamp) vm
+         return (caller', target, cd, num ts + timepassed)
+ let opts' = opts { testParams = testParams {testAddress = target, testCaller = caller', testTimestamp = timestamp'}}
+     thisCallRLP = List [BS $ word160Bytes caller', BS $ word160Bytes target, BS cd, BS $ word256Bytes timestamp']
+ -- set the timestamp
+ Stepper.evm $ assign (block . timestamp) (w256lit timestamp')
+ -- perform the call
+ bailed <- exploreStep opts' cd
+ Stepper.evm popTrace
+ let newHistory = if bailed then List history else List (thisCallRLP:history)
+     opts'' = opts {testParams = testParams {testTimestamp = timestamp'}}
+     carryOn = explorationStepper opts'' testName replayData newHistory (i - 1)
+ -- if we didn't revert, run the test function
+ if bailed
+ then carryOn
```

Comment:
Shouldn't we mark the test as failed if a call reverted? It's quite of a big semantic difference between the existing test types. If we're going to keep this behaviour perhaps we can highlight any reverts in the test output somehow?

Reply:
keep in mind that this is the "exploration calls" i.e. calls we're making to attempt to change state in some valid way. If it reverts, then there's no state change for us to care about